modules/pam_timestamp/hmac_openssl_wrapper.c

```diff
@@ -0,0 +1,405 @@
+/* Wrapper for hmac openssl implementation.
+ *
+ * Copyright (c) 2021 Red Hat, Inc.
+ * Written by Iker Pedrosa <ipedrosa@redhat.com>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, and the entire permission notice in its entirety,
+ *    including the disclaimer of warranties.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. The name of the author may not be used to endorse or promote
+ *    products derived from this software without specific prior
+ *    written permission.
+ *
+ * ALTERNATIVELY, this product may be distributed under the terms of
+ * the GNU Public License, in which case the provisions of the GPL are
+ * required INSTEAD OF the above restrictions.  (This clause is
+ * necessary due to a potential bad interaction between the GPL and
+ * the restrictions contained in a BSD-style copyright.)
+ *
+ * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED
+ * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#include "config.h"
+
+#ifdef WITH_OPENSSL
+
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <syslog.h>
+#include <unistd.h>
+#include <string.h>
+#include <errno.h>
+#include <openssl/hmac.h>
+
+#include <security/pam_ext.h>
+#include <security/pam_modutil.h>
+
+#include "hmac_openssl_wrapper.h"
+
+#define LOGIN_DEFS          "/etc/login.defs"
+#define CRYPTO_KEY          "HMAC_CRYPTO_ALGO"
```

Comment:
I am not really sure all this hmac algorithm configurability is needed here. If SHA2 based algorithms are broken then you do not have replacement among those supported here anyway. And if you really want to have algorithm agility, instead of hardcoding a table of algorithm names, you should use something like EVP_getdigestbyname() function to get the EVP_MD.

Also did you agree with shadow utils maintainer on addition of the new key value in login.defs?

Reply:
> I am not really sure all this hmac algorithm configurability is needed here. If SHA2 based algorithms are broken then you do not have replacement among those supported here anyway. And if you really want to have algorithm agility, instead of hardcoding a table of algorithm names, you should use something like EVP_getdigestbyname() function to get the EVP_MD.

Awesome! I didn't know about this function. Thanks for pointing it out to me.

> Also did you agree with shadow utils maintainer on addition of the new key value in login.defs?

No, and this is something that I wanted to discuss but somehow I forgot about it. I doubt what's the best approach to this configuration. On the one hand there's the current approach. On the other hand it would be possible to create a `pam_timestamp.conf` file and include it there. Finally, it would be possible to reuse current `ENCRYPT_METHOD` in `login.defs`. What do you think is the best solution?

Reply:
Can I get some feedback for this point? Every time I think about it I'm more willing to reuse `ENCRYPT_METHOD`as this  would mean changing only a single setting to improve the overall security of the system.

Reply:
The problem with ENCRYPT_METHOD potentially might be that there might be other values possible for ENCRYPT_METHOD in future that are not valid hash algorithm names. So basically you'd have to have some sensible default and if the ENCRYPT_METHOD fails to return valid hash algorithm then you'd have to fallback to the default.

Reply:
Thank you for the analysis. Then I think I'll continue with the current approach, so I'll contact shadow-utils maintainers to get their opinion.

Reply:
The key is available in shadow-utils master branch as https://github.com/shadow-maint/shadow/pull/333 got merged.